src/dart/lib/decoder.dart

```diff
@@ -0,0 +1,894 @@
+    final int blue = ((color.value & 0x000000FF) * alpha) ~/ 255;
+    return (red << 24) + (green << 16) + (blue << 8) + alpha;
+  }
+  
+  static Color RGBAToColor(int color) {
+    final int alpha = color & 0x000000FF;
+    if (alpha == 0x00) {
+      // We sometimes (for gradients) add in real color to the zero-alpha colors.
+      // These don't need to be unmultiplied further, they're already unmultiplied.
+      return Color((color & 0xFFFFFF00) >> 8);
+    }
+    assert(!isNonColor(color));
+    assert(((color & 0xFF000000) >> 24) <= alpha);
+    final int red = (((color & 0xFF000000) >> 24) * 255) ~/ alpha;
+    assert(((color & 0x00FF0000) >> 16) <= alpha);
+    final int green = (((color & 0x00FF0000) >> 16) * 255) ~/ alpha;
+    assert(((color & 0x0000FF00) >> 8) <= alpha);
+    final int blue = (((color & 0x0000FF00) >> 8) * 255) ~/ alpha;
+    return Color((alpha << 24) + (red << 16) + (green << 8) + blue);
+  }
+  
+  static const int kCREGLength = 64;
+  static const int kNREGLength = 64;
+  
+  final Uint32List customPalette = Uint32List.fromList(List<int>.filled(kCREGLength, 0x000000FF));
+  bool foundPalette = false;
+
+  final Uint32List CREG = Uint32List.fromList(List<int>.filled(kCREGLength, 0x000000FF));
+  final Float32List NREG = Float32List(kNREGLength);
+  late final double H;
+
+  int CSEL = 0;
+  int NSEL = 0;
+
+  double LOD0 = 0.0;
+  double LOD1 = double.infinity;
+
+  static bool isNonColor(int color) {
+    return (((color & 0xFF000000) >> 24) > (color & 0x000000FF))
+        || (((color & 0x00FF0000) >> 16) > (color & 0x000000FF))
+        || (((color & 0x0000FF00) >> 8) > (color & 0x000000FF));
+  }
+  static bool isGradient(int color) => (color & 0x000080FF) == 0x00008000;
+  
+  Gradient RGBAToGradient(int color, { ValueChanged<String>? onDebugDescribe }) {
+    assert(isGradient(color));
+    int NSTOPS = (color >> 24) & 0x3F;
+    final int CBASE = (color >> 16) & 0x3F;
+    final TileMode tileMode;
+    switch (color & 0x00C00000) {
+      case 0x00000000: // None
+        tileMode = TileMode.decal;
+        break;
+      case 0x00400000: // Pad
+        tileMode = TileMode.clamp;
+        break;
+      case 0x00800000: // Reflect
+        tileMode = TileMode.mirror;
+        break;
+      case 0x00C00000: // Repeat
+        tileMode = TileMode.repeated;
+        break;
+      default:
+        throw StateError('unreachable');
+    }
+    final int NBASE = (color >> 8) & 0x3F;
+    final bool isRadial = (color & 0x00004000) == 0x00004000;
+    final List<int> rawColors = List<int>.generate(NSTOPS, (int index) => CREG[(CBASE + index) % kCREGLength]).toList(); // "Color stops" in IconVG
+    final List<double> colorStops = List<double>.generate(NSTOPS, (int index) => NREG[(NBASE + index) % kNREGLength]); // "Offset stops" in IconVG
+
+    // IconVG uses premultiplied alpha, which means that it loses the information required to do gradients through a fully-transparent color.
```

Comment:
It's weirder that Flutter doesn't use premul alpha here - and if we ever "fix" that we should update this (https://github.com/flutter/flutter/issues/48674).

Reply:
While implementing this I was thinking the opposite -- that I wish this used straight alpha. :-)

Premultiplied alpha is less expressive.

Reply:
FWIW, in what I'd expect to be a 'fade out' common case, a gradient from `(R, G, B, A)` to `(R, G, B, 0)` in nonpremul space is equivalent to a gradient from `(R, G, B, A)` to `(0, 0, 0, 0)` in premul space. (I think; you might need to check my math).

For lerping from opaque pink to zero-alpha orange (not zero-alpha pink) then, yes, premul alpha is less expressive. But you might be able to approximate it by inserting multiple intermediate stops, including an ε-alpha orange just before the `(0, 0, 0, 0)` stop...